Release script for downstream OTP forks

t2gran opened the pull request:
### Summary

This PR contain two scripts `custom-release.py` and `changelog-diff.py` to make a release in a DOWNSTREAM fork of OTP. These scripts are not intended for releasing the main OTP repository.

See the (CUSTOM_RELEASE_README)[script/CUSTOM_RELEASE_README.md] for documentation of  `custom-release.py`.

The `changelog-diff.py` can be used to diff the _changelog.md_ file for any two versions of OTP.


### Issue

🟥  There is not issue for this. But, these scripts is requested from time to time. This PR provides a common place to maintain them. At Entur we now run the `custom-release.py` as the first step in our OTP CI/CD Pipeline.


### Unit tests

🟥  No tests provided.

### Documentation

✅  The scripts are documented in the readme, se above.

### Changelog

✅  This is relevant for advanced deplyments.


### Bumping the serialization version id

🟥  Not relevant.

t2gran commented:
I will add a check on the environment to the `prepare_release` - that the git repo is named correct. Other than that this is ready for review.

optionsome commented:
There are formatting issues.

t2gran commented:
I have improved the error handling in the scripts and simplified a bit. Ready for review.

optionsome commented on script/RELEASE_README.md:
```diff
@@ -0,0 +1,187 @@
+Also note that the CI configuration is changed (c2) and this change needs to be included in the
+new release (v3). 
+ 
+> **Note!** OTP ignore config it does not understand it. This allows us to roll out config for new 
```
```suggestion
> **Note!** OTP ignores config it does not understand it. This allows us to roll out config for new 
```

optionsome commented on script/RELEASE_README.md:
```diff
@@ -0,0 +1,187 @@
+
+You should create a branch in the local git repository where you keep your deployment-specific 
+config. The `ext_config` branch should have **one** commit with all changes needed. If you need to
+change the config, you will have to amend the changes. This is because the release script chery-pick
```
```suggestion
change the config, you will have to amend the changes. This is because the release script cherry-pick
```

optionsome commented on script/RELEASE_README.md:
```diff
@@ -0,0 +1,187 @@
+Run the `prepare_release` script. The script does the following:
+- Reset main to the right latest commit on `dev-2.x`
+- Rebase and merge in the `ext_config` extension branches into the local release branch. You will
+  be prompted for each step allowing you to perform/skip the steps you want. Also. the script keeps
```
```suggestion
  be prompted for each step allowing you to perform/skip the steps you want. Also, the script keeps
```

optionsome commented on script/prepare_release:
Shouldn't this file have `.sh` file extension?

optionsome commented on script/release:
Same as above.

optionsome reviewed (commented):
I would prefer if there was a directory under the script directory for these files so you can have just a regular README for it.

t2gran commented:
At Entur we use this this in the [GitHub workflow](https://github.com/entur/OpenTripPlanner/actions/workflows/entur-0-main-pipeline.yml). You may copy the workflow scripts if you want. Here is how we want to set up our pipeline - we have a data pipeline as well, but that is not shown.


![OTP GitHub Release Pipline](https://github.com/user-attachments/assets/a276ca27-b885-4939-b3d9-183cb043382d)

tkalvas commented on script/CUSTOM_RELEASE_README.md:
```diff
@@ -0,0 +1,240 @@
+> will regenerate the correct documentation file. After the test is run and new documentation is
+> generated you mark the conflict as resolved.
+
+If a conflic happens in the CI/CD pipline it is recomended to fix the branch causing the conflict.
```
conflic -> conflict

tkalvas commented on script/CUSTOM_RELEASE_README.md:
```diff
@@ -0,0 +1,240 @@
+> generated you mark the conflict as resolved.
+
+If a conflic happens in the CI/CD pipline it is recomended to fix the branch causing the conflict.
+The conflict can normally be fixed by rebasing or merging the extention branches or PRs. When 
```
extention -> extension

tkalvas commented on script/CUSTOM_RELEASE_README.md:
```diff
@@ -0,0 +1,240 @@
+
+## How-to print a changelog summary
+
+To add a summary in the CI/CD Pipline you can run the script standing at the tagged version you
```
Pipline -> Pipeline

tkalvas commented on script/custom-release.py:
```diff
@@ -0,0 +1,801 @@
+## ------------------------------------------------------------------------------------ ##
+
+POM_FILE_NAME = 'pom.xml'
+# GitHub label to indicate that a PR needs to be bumped
```
a PR -> the serialization id?

tkalvas commented on script/custom-release.py:
```diff
@@ -0,0 +1,801 @@
+# Configuration read from the 'custom-release-env.json' file
+class Config:
+    def __init__(self):
+        self.upstream_remote = None
```
It's called upstream_remote and base_revision, which... I guess it makes sense. One part of me would want to have them be either upstream_remote and upstream_revision, or base_remote and base_revision, to keep it consistent. Maybe this is a non-issue.

tkalvas commented on script/custom-release.py:
```diff
@@ -0,0 +1,801 @@
+
+# Make sure all remote repos are up-to date
+def fetch_all_git_remotes():
+    git('fetch', '--all', error_msg='Git fetch all remotes failed!')
```
This is wildly improbable, but someone might have extra remotes defined, some of which are not required for the release to succeed. If they happen to be unavailable for any reason, this fetch will make the script fail and exit, even though it could have continued. Not worth changing I think, but maybe a comment about it for the future person who hits the problem.

tkalvas commented on script/custom-release.py:
```diff
@@ -0,0 +1,801 @@
+
+def resolve_new_version():
+    info('Resolve new version number ...')
+    p = git('tag', '--list', '--sort=-v:refname', error_msg='Fetch git tags failed!')
```
The --sort is unnecessary, because line 439 max_tag_version = max(<generator>) goes through the whole list anyway.

tkalvas commented on script/custom-release.py:
```diff
@@ -0,0 +1,801 @@
+        return
+    info('Get PRs to include and their labels from GitHub - This requires authentication ...')
+
+    # The query body needs to be on one line, for an unknown reason.
```
Would still be nicer to read if the query was in the script with newlines and then query_text = re.sub(r'\n\s*', ' ', query_text)

t2gran replied:
I changed it. I used a slightly different REGEXP: `re.sub(r'(?s)\s{2,}', ' ', query_text)`, this 
replace 2 or more whitespace with a singe space including newline, formfeed and tab.

tkalvas commented on script/custom-release.py:
```diff
@@ -0,0 +1,801 @@
+        config.ser_ver_id_prefix = doc['ser_ver_id_prefix']
+        debug(f'Config loaded: {config}')
+
+    if len(config.ser_ver_id_prefix) != 2:
```
Checks the prefix is exactly two characters long, documentation says it is a maximum of two characters long.

tkalvas commented on script/CUSTOM_RELEASE_README.md:
```diff
@@ -0,0 +1,240 @@
+  "upstream_remote": "otp",
+  "release_remote": "<organization name>",
+  "release_branch": "<release branch>",
+  "ser_ver_id_prefix": "<organization abbrivation, max 2 characters>",
```
Says max 2 characters, but code checks and relies on it being exactly 2 characters.

tkalvas commented on script/custom-release.py:
```diff
@@ -0,0 +1,801 @@
+
+
+def bump_release_ser_ver_id(current_id):
+    value = int(current_id[3:])
```
Skips 3 characters which means the ser_ver_id_prefix must be exactly 2 characters long (which is checked, but documentation claims it is a maximum of 2 characters long).

tkalvas commented on script/custom-release.py:
```diff
@@ -0,0 +1,801 @@
+        print(f'DEBUG {msg}', flush=True)
+
+
+def crop(text):
```
Someone will come up with an error situation where they want the entire output.

t2gran replied:
Improvements are always welcome. For example dumping everything to a `custom-release.log` file and crop terminal output is fine with me. The script potentially print the 20 versions of the root `pom.xml` in addition to running Maven, so the log can be long - I prefer cropping some of it, so it is easier to find the important stuff. I have also modified the script to print everything a couple of times to debug.

t2gran replied:
Thank you for a good review @tkalvas, I renamed and rewrote a lot to make it generic (not Entur specific) - sorry for all the forgotten updates. 

I have tested the script after my changes - it still works.

tkalvas replied:
Yes, I guess it's best to crop the log which is going to stdout, and write the full log to a file in /tmp

t2gran commented on script/custom-release.py:
```diff
@@ -0,0 +1,801 @@
+    See the RELEASE_README.md for more details.
+
+    Usage
+      script/prepare_release.py [options] [<base-revision>]
```
Fiexed!

t2gran commented on script/custom-release.py:
```diff
@@ -0,0 +1,801 @@
+      --summary  : Print a markdown summary to '.custom_release_summary.md'
+
+    Examples
+      # script/prepare_release.py otp/dev-2.x --printSummary
```
Fiexed in all places.

t2gran commented on script/CUSTOM_RELEASE_README.md:
```diff
@@ -0,0 +1,240 @@
+# README - OTP Cusom/Fork Release Scripts
```
fixed

t2gran commented:
Thank you for a good review @tkalvas, I renamed and rewrote a lot to make it generic (not Entur specific) - sorry for all the forgotten updates. 

I have tested the script after my changes - it still works.
